sklearn/tests/test_metadata_routing.py

```diff
@@ -844,3 +848,135 @@ def test_metadata_routing_get_param_names():
+            return self.estimator_.predict(X, **routed_params.estimator.predict)
+
+        def get_metadata_routing(self):
+            return (
+                MetadataRouter(owner=self.__class__.__name__)
+                .add(estimator=self.estimator, method_mapping="one-to-one")
+                .warn_on(
+                    child="estimator",
+                    method="fit",
+                    params=None,
+                    raise_on="1.4",
+                )
+            )
+
+    class Estimator(BaseEstimator):
+        def fit(self, X, y, sample_weight=None, groups=None):
+            return self
+
+        def predict(self, X, sample_weight=None):
+            return np.ones(shape=len(X))
+
+    est = MetaEstimator(estimator=Estimator())
+    # the meta-estimator has est to have a warning on `fit`.
+    with pytest.warns(
+        FutureWarning, match="From version 1.4 this results in the following error"
+    ):
+        est.fit(X, y, sample_weight=my_weights)
+
+    # but predict should raise since there is no warn_on set for it.
+    with pytest.raises(
+        ValueError,
+        match=re.escape(
+            "[sample_weight] are passed but is not explicitly set as requested or not"
+            " for Estimator.predict"
+        ),
+    ):
+        est.predict(X, sample_weight=my_weights)
+
+    # In this case both a warning and an error are raised. The warning comes
+    # from the MetaEstimator, and the error from WeightedMetaRegressor since it
+    # doesn't have any warn_on set but sample_weight is passed.
+    est = MetaEstimator(estimator=WeightedMetaRegressor(estimator=RegressorMetadata()))
+    with pytest.raises(
+        ValueError,
+        match=re.escape(
+            "[sample_weight] are passed but is not explicitly set as requested or not"
+            " for RegressorMetadata.fit"
+        ),
+    ):
+        with pytest.warns(
+            FutureWarning, match="From version 1.4 this results in the following error"
+        ):
+            est.fit(X, y, sample_weight=my_weights)
+
+    class WarningWeightedMetaRegressor(WeightedMetaRegressor):
+        """A WeightedMetaRegressor which warns instead."""
+
+        def get_metadata_routing(self):
+            router = (
+                MetadataRouter(owner=self.__class__.__name__)
+                .add_self(self)
+                .add(estimator=self.estimator, method_mapping="one-to-one")
+                .warn_on(
+                    child="estimator",
+                    method="fit",
+                    params=["sample_weight"],
+                    raise_on="1.4",
+                )
+                .warn_on(
+                    child="estimator",
+                    method="score",
+                    params=["sample_weight"],
+                    raise_on="1.4",
+                )
+            )
+            return router
+
+    # Now there's only a warning since both meta-estimators warn.
```

Comment:
I would split into 2 tests. One that checks that we raise an error.
The other one checking that we raise a warning.